sklearn/naive_bayes.py

```diff
@@ -117,6 +118,303 @@ def predict_proba(self, X):
+            Estimator must support :term:`fit`, :term:`predict`
+            and :term:`_joint_log_likelihood`.
+        column(s) : array-like of string or int, slice
+            Indexes the data on its second axis. Integers are interpreted as
+            positional columns, while strings can reference DataFrame columns
+            by name. A scalar string or int should be used where
+            ``transformer`` expects X to be a 1d array-like (vector),
+            otherwise a 2d array will be passed to the transformer.
+
+    Attributes
+    ----------
+    classes_ : ndarray of shape (n_classes,)
+        class labels known to the classifier
+
+    n_features_ : int
+        Number of features of each sample.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import pandas as pd
+    >>> from sklearn.naive_bayes import GeneralNB, GaussianNB, BernoulliNB
+
+    >>> X = np.array([[1.5, 2.3, 5.7, 0, 1],
+    ...               [2.7, 3.8, 2.3, 1, 0],
+    ...               [1.7, 0.1, 4.5, 1, 0]])
+    >>> y = np.array([1, 0, 0])
+    >>> X_test = np.array([[1.5, 2.3, 5.7, 0, 1]])
+    
+    >>> clf = GeneralNB([
+    >>>     ("gaussian", GaussianNB(), [0, 1, 2]),
+    >>>     ("bernoulli", BernoulliNB(), [3, 4])
+    >>> ])
+    >>> clf.fit(X, y)
+    >>> print(clf.predict(X_test))
+    [1]
+    >>> print(clf.score([[2.7, 3.8, 1, 0, 1]],[0]))
+    1.0
+
+    >>> df = pd.DataFrame(X)
+    >>> df.columns = list("abcde")
+    >>> df["y"] = [1, 0, 0]
+    >>> clf = GeneralNB([
+    >>>     ("gaussian", GaussianNB(), ["a", "b", "c"]),
+    >>>     ("bernoulli", BernoulliNB(), ["d", "e"])
+    >>> ])
+    >>> clf.fit(df.iloc[:,:-1], df["y"])
+
+    >>> df_test = pd.DataFrame(X_test)
+    >>> df_test.columns = list("abcde")
+    >>> print(clf.predict(df_test))
+    [1]
+    >>> print(clf.score([[2.7, 3.8, 1, 0, 1]],[0]))
+    1.0
+    """
+    # TODO consider jll for each estimator
+    # TODO unify variable names with similar meaning 
+    #      ("distribution", "model", "nb"), ("column", "feature")
+
+    def __init__(self, distributions):
+        self.distributions = distributions
+        self.classes_ = None
+        self.n_features_ = None
+        self._fits = None
+        self._columns = None
+        self._df_columns = None
+        self._is_fitted = False
+
+    def fit(self, X, y):
+        """Fit Gaussian Naive Bayes according to X, y
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+        self._validate_distributions(X)
+        self._check_X_y(X, y)
+
+        self.classes_ = np.unique(y)
+
+        self._fits = [
+            (nb.fit(_safe_indexing(X, features, axis=1), y), features)
```

Comment:
if features is a callable, this won't work. This needs testing. It is useful to allow a callable `features`, allowing for technologies like `sklearn.compose.make_column_selector`. As in ColumnTransformer, a callable `features` needs to be computed at `fit` time and reused later.

Reply:
Let me read up on this a little bit more.

Reply:
I've now incorporated preprocessing the callables like the `sklearn.compose.make_column_selector` API.